#6 회원 가입 API

donggi-lee-bit opened the pull request:
안녕하세요 @f-lab-maverick 코드가 포함된 첫 리뷰 요청입니다! 잘 부탁드리겠습니다 🥳
MyBatis 사용이 처음이라 동작하는 코드를 만드는데 애를 좀 먹었습니다.. ㅎㅎ

### 고려한 것
- 최소한의 테스트 코드를 작성하려고 했습니다.
  - 도메인 단위 테스트
  - API 통합 테스트
  - repository를 통한 데이터 저장, 비밀번호 암호화 적용과 같은 특정 메서드 테스트를 위한 슬라이스 테스트
- Validator 유틸 클래스를 추가하여 사용자의 입력 값과 도메인의 필드 값을 검증하였습니다. 
  - 이때 데이터가 들어오는 시점과 저장 시점 두 번 검증하는 방식이 괜찮은가 하는 의문이 있습니다.
  - 비밀번호의 경우 raw 값과 encode 값이 다르기 때문에 raw 값의 길이와 encode 길이를 각각 확인해야하지만, null 체크는 2번 할 필요가 있을까, 1번을 한다면 어디서 하는 게 좋을까 하는 의문이 들었습니다.

### 테스트 코드 결과
<img width="536" alt="image" src="https://github.com/user-attachments/assets/b135102b-d3e6-4dfd-8614-866364359e1d">

donggi-lee-bit committed:
패키지에서 언더바를 제거하기 위해 네이밍 변경

donggi-lee-bit committed:
멤버 등록 해피 케이스 테스트 코드 추가

donggi-lee-bit committed:
멤버 등록 시 email 필드에 대한 예외 케이스 추가

donggi-lee-bit committed:
멤버 등록 시 password 필드에 대한 예외 케이스 추가

donggi-lee-bit committed:
도메인 테스트 내에 DTO 제거

donggi-lee-bit committed:
테스트 코드 중복 계층 제거

donggi-lee-bit committed:
Validator 유틸 클래스 및 커스텀 예외 클래스 추가

donggi-lee-bit committed:
MyBatis를 사용해 Member 등록

donggi-lee-bit committed:
회원 가입 API 추가

donggi-lee-bit committed:
회원 가입 시 이메일이 중복될 경우 409 에러 응답 테스트

donggi-lee-bit committed:
회원 가입 시 클라이언트의 입력 값이 유효하지 않을 때 에러 응답 테스트

donggi-lee-bit committed:
회원 가입 시 비밀번호를 암호화하여 저장

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/config/SecurityConfig.java:
```diff
@@ -0,0 +1,30 @@
+@Configuration
+public class SecurityConfig {
+
+    @Bean
```
Spring boot를 사용하면서 관용적으로 사용하는 어노테이션들이 꽤나 많습니다.
이런 어노테이션들이 어떤 역할을 하는 것인지 주석으로 적어보면 어떨까요? 특히 이런 어노테이션은 공식문서에 상세하게 설명되어있기 때문에 찾아보시면 도움이 많이 됩니다.

예시: https://docs.spring.io/spring-framework/reference/core/beans/java/configuration-annotation.html#page-title

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/config/SecurityConfig.java:
```diff
@@ -0,0 +1,30 @@
+public class SecurityConfig {
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
```
BCryptPasswordEncoder를 기본 생성자를 이용하여 생성하면 어떻게 동작할까요?
커스텀할 수 있는 방법은 어떤게 있을까요?

donggi-lee-bit replied:
`BCryptPasswordEncoder`를 기본 생성자로 생성할 경우 기본 strength 값이 **10**으로 설정됩니다. strength는 해싱 알고리즘의 연산 횟수로, 값이 클수록 보안성이 강화되지만, 해싱 속도가 느려진다는 단점이 있습니다.

`BCryptPasswordEncoder`를 커스텀 하는 방법에는

- strength 값을 조정하여 연산 횟수를 늘리거나 줄일 수 있습니다.
- SecureRandom 인스턴스를 추가하여 난수 생성 방식을 설정할 수 있습니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/config/SecurityConfig.java:
```diff
@@ -0,0 +1,30 @@
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+            .csrf(AbstractHttpConfigurer::disable)
```
API서버에서는 csrf를 disable처리하곤 합니다. csrf란 무엇이며, disable할 경우 어떤 일이 발생할까요?

donggi-lee-bit replied:
CSRF(cross-site request forgery)는 공격자가 사용자의 권한을 악용해 사용자가 의도하지 않은 요청을 서버로 전송하도록 유도하는 웹 공격 방식입니다. 이를 통해 공격자는 계좌 이체, 비밀번호 변경 등 사용자의 민감한 작업을 수행할 수 있습니다.

API 서버에서 CSRF를 disable하게 될 경우, 쿠키 기반 인증을 사용할 경우 CSRF 공격에 노출될 수 있습니다. 하지만 클라이언트가 요청마다 `Authorization` 헤더를 사용해 인증 정보를 명시하여 전달할 경우, csrf 공격에 노출될 수 있는 쿠키 기반 인증의 위험성이 제거되므로 상대적으로 안전합니다.

또한 이를 보완하기 위해, CORS 정책을 설정하여 신뢰할 수 있는 출처에서만 요청을 허용하고, HTTPS를 사용하여 요청과 응답 데이터를 암호화하여 안전한 통신이 될 수 있게 해야합니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/config/SecurityConfig.java:
```diff
@@ -0,0 +1,30 @@
+        http
+            .csrf(AbstractHttpConfigurer::disable)
+            .authorizeHttpRequests((auth) -> auth
+                .requestMatchers("/v1/**").permitAll()
```
`requestMatchers` 이외에도 다양한 matcher가 있습니다.
`requestMatchers`는 어떤방식으로 매칭할까요? 특정 HTTP Method에 대해서만 권한을 설정하려면 어떤 Matcher를 사용해야할까요?

donggi-lee-bit replied:
`requestMatchers` 는 기본적으로 `AntPathMatcher` 를 사용하여 요청 URL을 매칭합니다. 예를 들어 `/api/**`, `/v1/**` 와 같은 패턴을 지원합니다. 그 외에도 특정 HTTP 메서드, 헤더, 파라미터, Content-Type 등에 대해 조건을 추가하여 매칭할 수 있습니다.

또한 `MvcRequestMatcher`, `RegexRequestMatcher` 등 다른 Matcher를 사용하여 다양한 형태의 요청 URL을 매칭할 수 있도록 확장이 가능합니다.

특정 HTTP Method에 대해서만 권한을 설정하려면 `requestMatchers`  메서드를 이용하여 HTTP 메서드와 URL 패턴을 함께 지정할 수 있습니다.
    
```java
http
    .authorizeHttpRequests((auth) -> auth
        .requestMatchers(HTTPMethod.POST, "/v1/**").permitAll()
```

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/utils/RegexPattern.java:
```diff
@@ -0,0 +1,22 @@
+package com.donggi.sendzy.common.utils;
+
+public enum RegexPattern {
+    EMAIL("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,6}$", "유효한 이메일 형식 (예: user@example.com)"),
+    PASSWORD("^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[!@#$%^&*]).{8,}$", "영문 대문자, 영문 소문자, 숫자, 특수문자 중 3가지 이상 포함");
```
특정 정규식을 저장하기 위해 enum을 쓰셨네요.
- enum을 쓴 이유가 있나요? 
- enum의 어떤 기능을 위해 이렇게 작성하셨나요?
- 상수로 선언하는것과는 어떤 차이가 있을까요?

donggi-lee-bit replied:
해당 정규식은 현재 Model 클래스와 회원 가입 요청 DTO 클래스에서 사용하고 있습니다. 비밀번호 수정이나 이메일 수정 기능이 추가된다면, 수정 요청 DTO 클래스에서도 동일한 정규식이 사용하게 될 것입니다.

정규식을 각 클래스에서 상수로 관리하게 되면, 정규식이 변경되었을 때 각 클래스에 선언된 상수를 각각 수정해줘야 하므로 변경 작업이 누락될 가능성이 있습니다. 이를 방지하고 코드의 일관성을 유지하기 위해 enum 에서 정규식을 관리하게 하였습니다.

f-lab-maverick replied:
@donggi-lee-bit 
논리적으로 잘 설명해주셨지만 왜 Enum을 사용했는지에 대한 타당성이 부족합니다. enum의 특징인 타입안정성, 집합 관리 등의 기능이 필요한지 / 상수와의 비교를 했을 때 어떤 차이가 있는지를 보강해주세요. static final 상수로 관리하면 말씀해주신 "정규식이 변경되었을 때 각 클래스에 선언된 상수를 각각 수정해줘야" 하는 단점이 동일하게 보강됩니다. 답변을 더 보강해주세요~

donggi-lee-bit replied:
말씀하신대로 정규식을 `static final` 상수의 형태로 관리해도 코드의 일관성은 유지될 수 있습니다. 그래서 정규식만 관리한다면 `static final` 상수로 관리해도 충분하다고 생각합니다. 

하지만 여기에 정규식의 메타 정보를 함께 관리하기 위해서는 enum이 더 적합하다고 판단했습니다. Enum을 사용하면 정규식과 메타 정보를 하나의 데이터 집합으로 통합하여 관리할 수 있어 코드의 가독성과 유지보수성 측면에서 더 유리하다고 생각합니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/utils/RegexPattern.java:
```diff
@@ -0,0 +1,22 @@
+    public String getDescription() {
+        return description;
+    }
+}
```
enum에도 Lombok의 `@Getter`를 사용할 수 있습니다. 이런 관용적인 메서드는 직접 작성할 필요는 없습니다~

f-lab-maverick commented on src/main/java/com/donggi/sendzy/common/utils/Validator.java:
```diff
@@ -0,0 +1,64 @@
+        if (input == null || input.isBlank()) {
+            throw new ValidException("%s 은/는 null 또는 공백이 될 수 없습니다.".formatted(fieldName));
+        }
+    }
```
직접 Validator를 구현하셨네요~ 사실 이렇게 공용으로 사용하는 유효성검증은 Spring의 Validation 모듈을 사용하면 직접 구현하지 않아도 처리할 수 있습니다. 다만 직접 구현할경우 테스트코드를 작성하기는 좋다는 장점이 있겠네요!

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/application/SignupService.java:
```diff
@@ -0,0 +1,29 @@
+    public SignupService(final MemberService memberService, final PasswordEncoder passwordEncoder) {
+        this.memberService = memberService;
+        this.passwordEncoder = passwordEncoder;
+    }
```
Lombok의 `@RequiredArgsConstructor`를 사용하면 관습적인 생성자 코드를 제거할 수 있습니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/application/SignupService.java:
```diff
@@ -0,0 +1,29 @@
+        }
+
+        memberService.save(request.email(), passwordEncoder.encode(request.password()));
+    }
```
이 메서드는 동시성이슈의 발생 가능성이 있습니다.

1. 같은 이메일로 가입한 회원이 없음을 확인한다.
2. 회원을 저장한다

두가지 액션이 이루어지기 때문에 race condition이 발생할 경우 동일한 이메일의 회원이 여러명 가입될 수 있습니다.
이를 어떻게 방지할 수 있을까요?

donggi-lee-bit replied:
먼저 이메일 컬럼에 `unique` 제약 조건을 걸어서 동일한 이메일로 여러 명의 회원이 가입되지 않도록 하였습니다. 

이 경우 existsByEmail() 메서드에서 중복이 잡히지 않더라도, save 시점에서 `DataIntegrityViolationException`이 발생할 수 있습니다. 해당 예외를 `ControllerAdvice` 에서 잡아서 사용자에게는 시스템에 문제가 있다는 메세지를 출력하는 정도로 마무리할 것 같습니다.

f-lab-maverick replied:
DB의 unique 제약조건을 거는것도 좋은 방법이죠~ 사실 다른 방식도 꽤나 많이 있지만, 이 부분은 추후 유니크제약조건으로 해결되지 않는 비즈니스로직을 만났을 때 개선해보도록 하죠.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/application/SignupService.java:
```diff
@@ -0,0 +1,29 @@
+            throw new EmailDuplicatedException("이미 가입된 이메일입니다.");
+        }
+
+        memberService.save(request.email(), passwordEncoder.encode(request.password()));
```
특정 테이블을 저장할 때는 해당 테이블의 데이터를 모두 파라미터로 받기보다는 Model 클래스를 생성해주는 것이 좋습니다. 이렇게 하면 코드의 가독성과 유지보수성이 향상됩니다. 예를 들어, request.email()과 passwordEncoder.encode(request.password())와 같은 형태 대신, 하나의 Member 객체를 생성하여 이를 서비스 레이어로 전달하면 코드가 훨씬 간결하고 명확해집니다.

다음과 같은 방식으로 리팩토링할 수 있습니다

```
Member member = new Member(request.email(), passwordEncoder.encode(request.password()));
memberService.save(member);
```

donggi-lee-bit replied:
말씀하신 방법대로 Model 클래스를 인자로 전달하면, Model 클래스에 필드가 추가되더라도 수정 범위가 줄어들고, 주어진 매개변수를 기반으로 Member 객체를 생성하여 저장한다는 흐름이 더 명확해질 것 같습니다. 제안해주신 방식대로 수정하겠습니다!

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/domain/Member.java:
```diff
@@ -0,0 +1,51 @@
+
+public class Member {
+
+    private static final int EMAIL_MAX_LENGTH = 100;
+    private static final int EMAIL_MIN_LENGTH = 5;
```
Length를 저장하는 별도 클래스를 만들변 매번 변수를 두개씩 만들 필요가 없지 않을까요? 
```
@RequiredArgsConstructor
class Length {
  private final int min;
  private final int max;
}

private static final Length EMAIL_LENGTH = new Length(5, 100);
...
```

donggi-lee-bit replied:
와 좋은 방법입니다!! 
여기에 Length 객체를 생성할 때 빌더 패턴을 적용하면 매개변수의 의미를 더 명확하게 표현 할 수 있을 거 같은데 어떤가요??

```java
@Getter
public class Length {

    private final int min;
    private final int max;

    private Length(final Builder builder) {
        this.min = builder.min;
        this.max = builder.max;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private int min;
        private int max;

        public Builder min(final int min) {
            this.min = min;
            return this;
        }

        public Builder max(final int max) {
            this.max = max;
            return this;
        }

        public Length build() {
            return new Length(this);
        }
    }
}

// builder 패턴을 사용한 Length 객체 생성 예시
private static final Length EMAIL_LENGTH = Length.builder().min(5).max(100).build();
private static final Length PASSWORD_LENGTH = Length.builder().min(8).max(100).build();
```

f-lab-maverick replied:
빌더패턴은 주의해서 사용해야합니다. 빌더 특성상 중간에 생성 매개변수를 빼먹기 아주 쉬운 구조입니다. 생성자의 경우 명시적으로 null을 넣어야 초기회가 안되지만, 빌더패턴은 메서드 체이닝을 빼먹으면 바로 NPE거든요.

따라서 이 경우 생성자를 사용하는 것이 더 안전하다고 볼 수 있겠습니다. 빌더는 Config 데이터를 설정할 때 처럼 대부분의 필드에 기본값이 있는 상태에서 데이터를 override할 때 적합합니다.

donggi-lee-bit replied:
NPE 문제가 생길 수 있다는 걸 놓쳤습니다. 수정하겠습니다!

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/domain/Member.java:
```diff
@@ -0,0 +1,51 @@
+
+    private Long id;
+    private String email;
+    private String password;
```
이 Member의 password는 이미 암호화된 값이 들어가기 때문에, 변수명을 `encryptedPassword` 로 지어주는게 덜헷갈립니다.
안그러면 이 패스워드가 암호화가 된건지 안된건지 알기가 어렵기때문에, 변수명은 헷갈릴 가능성을 없애도록 지어주는게 좋습니다.

암호화 되지 않은 평문은 `plainPassword` 등으로 변수명을 지어주면 더 좋겠죠?

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/domain/Member.java:
```diff
@@ -0,0 +1,51 @@
+        validate(email, password);
+        this.email = email;
+        this.password = password;
+        this.createdAt = LocalDateTime.now();
```
Java레벨에서 서버의 현재 시간을 입력하여 insert하는것과, insert문 내부에서 DB의 시간으로 `now()`를 호출하여 저장하는것은 어떤 차이가 있을까요?

EX)
1안 서버시간을 직렬화하여 직접 insert. `INSERT INTO MEMBER VALUES ('...', '...', '2024-10-10 10:00')`
2안. DB시간으로 저장. `INSERT INTO MEMBER VALUES ('...', '...', 'now()'`

donggi-lee-bit replied:
Java 레벨에서 서버의 현재 시간을 입력하여 insert 할 경우

- 시간 제어를 애플리케이션에서 하기 때문에 테스트 환경에서 커스텀한 시간 데이터를 만들어 적용할 수 있어 시간과 관련된 다양한 시나리오의 테스트에 용이합니다.
- DB가 변경되거나, 현재 시간을 가져오는 함수의 동작이 바뀌어도 일관된 시간 데이터를 저장할 수 있습니다.

f-lab-maverick replied:
다른 부분도 고려해볼 수 있습니다.
운영환경의 자바 어플리케이션은 가용성을 위해 2개 이상의 어플리케이션을 띄워 운용하는 경우가 대부분입니다. 이 경우 각 서버의 시스템 시간이 다를 수 있습니다. 그렇다면 동시에 들어간 데이터라도 `createdAt`이 큰 차이가 날 수 있습니다.

DB는 대부분 1개의 Master로 운영하기떄문에 `createdAt`의 타이밍 정보가 중요하다면 DB에서의 now()를 사용하는게 적절할 수도 있겠습니다. 물론 로그성 데이터라면 어느정도의 시차는 허용되기때문에 이슈 없습니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/domain/Member.java:
```diff
@@ -0,0 +1,51 @@
+        this.createdAt = LocalDateTime.now();
+    }
+
+    private void validate(final String email, final String password) {
+        validateEmail(email);
+        validatePassword(password);
+    }
```
이 메서드가 member에 존재할 필요가 있을까요? 특히 password의 경우 Member 클래스로 생성되었다면 이미 암호화가 되어있는 데이터가 넘어올 것입니다. 암호화가 완료된 패스워드를 validation하는게 의미가 있을까요?

donggi-lee-bit replied:
암호화된 패스워드는 validation할 필요가 없습니다. Member에 password를 먼저 평문으로 개발한 뒤 암호화 로직을 추가하면서 불필요한 검증 로직이 남게 되었습니다. 해당 validation 로직은 제거하도록 하겠습니다!

donggi-lee-bit replied:
이메일의 경우에는 Request DTO에서 검증하고 있으니 Member 클래스에서는 해줄 필요가 없을까요??

donggi-lee-bit replied:
말씀하신대로 password의 경우 이미 암호화된 데이터가 넘어오기 때문에 null, length, regex를 검사하는건 의미가 없으므로 제거되는 게 맞는 것 같습니다!

다만, Model 클래스 내부에 지금과 같은 validation 로직을 작성하는 것에 대해 궁금한 게 있습니다.

Model 클래스 내부에 validation 로직을 두면 문서화의 이점이 있을 거라고 생각했습니다. 필드에 대한 validation 로직이 클래스 내부에 위치함으로써, 코드만으로도 각 필드의 제약 조건을 명확히 알 수 있다는 장점이 있다고 생각했습니다. 이는 코드를 읽는 개발자 입장에서 클래스의 제약 조건을 빠르게 이해하는 데 도움이 되지 않을까? 하는 생각을 해봤습니다.

하지만 이 방식이 실제로 유용할 지 의문이 있기도 합니다. maverick은 이에 대해 어떻게 생각하시나요??

f-lab-maverick replied:
말씀하신 방식으로 만들려면 우선 두가지를 분리해야합니다.
1. 평문상태의 암호
2. 암호화된 암호

1, 2번은 명확하게 validation로직이 다릅니다. 따라서 두 클래스를 분리하고 각각의 validation을 모델 내부에 둔다면 적절한 제약사항 분리 + 코드 응집도 강화로 볼 수 있습니다. 말씀해주신대로 제약조건의 이해나 코드를 읽을 때 도움이 된다는 장점이 있습니다.

donggi-lee-bit replied:
평문 비밀번호와 암호화된 비밀번호의 validation 로직을 분리하도록 하겠습니다! 이 부분에 대해 예전부터 궁금했었는데, 이번 기회에 궁금증을 해소할 수 있었습니다. 감사합니다!

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/dto/SignupRequest.java:
```diff
@@ -0,0 +1,40 @@
+    String email,
+    String password
+) {
+    private static final int EMAIL_MAX_LENGTH = 100;
+    private static final int EMAIL_MIN_LENGTH = 5;
+    private static final int RAW_PASSWORD_MAX_LENGTH = 32;
+    private static final int RAW_PASSWORD_MIN_LENGTH = 5;
```
비슷한 validation 변수가 있군요. Request DTO와 Model의 validation이 다를 수 있으므로 클래스마다 이렇게 유효성검증 관련 데이터를 분리해둔건 좋은 방식입니다. 👍

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/dto/SignupRequest.java:
```diff
@@ -0,0 +1,40 @@
+import com.donggi.sendzy.common.utils.RegexPattern;
+import com.donggi.sendzy.common.utils.Validator;
+
+public record SignupRequest(
+    String email,
+    String password
+) {
```
record클래스를 사용하셨네요. record클래스는 장단점이 꽤 있는 방식입니다. 어떤 특징이 있을까요?

donggi-lee-bit replied:
record 클래스의 장점은
- 필드가 `final`로 선언되어 불변 객체를 보장합니다.
- 필드에 따라 생성자, `getter`, `equals()`, `hashcode()`, `toString()`을 자동으로 생성해주기 때문에 보일러 플레이트 코드가 줄어듭니다.

record 클래스의 단점은
- record 클래스는 `final`로 선언됩니다. 따라서 record 클래스는 확장이 불가능합니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/infrastructure/MemberMapper.java:
```diff
@@ -0,0 +1,21 @@
+import org.apache.ibatis.annotations.Mapper;
+import org.apache.ibatis.annotations.Select;
+
+@Mapper
+public interface MemberMapper extends MemberRepository {
+
+    @Insert("INSERT INTO member (email, password, created_at) VALUES (#{email}, #{password}, #{createdAt})")
+    int save(Member member);
+
+    @Select("SELECT EXISTS (SELECT 1 FROM member WHERE email = #{email})")
+    boolean existsByEmail(String email);
+
+    @Delete("DELETE FROM member")
+    void deleteAll();
+}
```
가능하면 쿼리는 xml파일 내부에 작성해주세요. 쿼리가 길어질경우 가독성이 떨어지기 때문에 우선 xml파일로 매핑하는 방식을 사용해주시면 좋겠습니다~

f-lab-maverick commented on src/test/java/com/donggi/sendzy/member/application/SignupServiceTest.java:
```diff
@@ -0,0 +1,55 @@
+        // then
+        verify(passwordEncoder).encode(rawPassword);
+        verify(memberService).save(email, encodedPassword);
+    }
```
테스트코드도 짜주셨네요~ 현재단계에서 테스트코드는 수정 범위에서는 제외해주셔도 좋습니다. JUnit관련 문서와 Spring Test 문서를 학습한 후 좀 더 깊게 들어가보도록 하겠습니다. 테스트 리뷰는 참고만 해주세요.

테스트를 짤 때에는 해당 테스트가 어떤 목적을 가지고있고, 성공조건이 무엇이며 이 테스트가 깨질경우 어떤 부분이 문제가되는지를 명확하게 파악할 수 있어야 합니다.

이 테스트의 경우 테스트가 깨진다면 제목만보고 어떤게 문제인지 알기는 적절합니다. 다만, Mock 프레임워크를 사용할경우 구현이 변경되거나 리팩토링이 있을 경우 동작이 동일하더라도 모든 테스트가 깨지기때문에 사용에 주의가 필요합니다.

따라서 가능하면 Mock을 사용하지 않은 채로 테스트를 작성하는 연습을 먼저 하는게 좋습니다.
이 경우 프레임워크레벨에서 제공해주는 기능을 테스트하기 어려우므로, POJO로 테스트할 수 있는 유틸성클래스를 대상으로 먼저 테스트코드 작성을 연습해주시면 됩니다.

f-lab-maverick commented on src/test/java/com/donggi/sendzy/member/application/SignupServiceTest.java:
```diff
@@ -0,0 +1,55 @@
+        MockitoAnnotations.openMocks(this);
+    }
+
+    @Test
+    void 비밀번호가_암호화된_상태로_저장된다() {
```
JUnit 5부터는 `@DisplayName`(https://junit.org/junit5/docs/current/user-guide/#writing-tests-display-names) 어노테이션을 제공하여 테스트이름을 자유롭게 문자열로 작성할 수 있습니다. 메서드명을 한글로 적는건 보통 JUnit4에서 하던 방식이고, 메서드명은 자유롭게 / DIsplayName에 명시적인 테스트명을 작성해주는게 가독성이 더 좋습니다.

donggi-lee-bit replied:
모든 테스트에 `@DisplayName`을 개별적으로 설정하는 번거로움을 줄이기 위해 테스트 클래스에 `@DisplayNameGenerator` 를 지정하고 테스트 의도를 메서드에 명시하였습니다. 테스트 코드를 이와 같이 작성하면 테스트의 의도를 메서드와 `@DisplayName` 양쪽 다 명시하지 않아도 되고, 메서드 이름을 기반으로 디스플레이 이름이 자동으로 생성되므로 테스트 작성 방식을 단순화할 수 있다고 생각합니다.

f-lab-maverick commented on src/test/java/com/donggi/sendzy/member/domain/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,31 @@
+    @Autowired
+    private MemberRepository memberRepository;
+
+    @Test
+    void 멤버를_저장하고_저장된_행의_수를_반환한다() {
+        // given
+        final var email = "donggi@sendzy.com";
+        final var password = "supperPass1word@#";
+
+        // when
+        final var actual = memberRepository.save(new Member(email, password));
+
+        // then
+        assertThat(actual).isEqualTo(1L);
+    }
```
이 테스트를 시작하는 시점에 누군가가 이미 "donggi@sendzy.com" 이메일로 가입했다면 어떤 일이 발생할까요?
비즈니스 로직이 변하지 않았고, 잘 동작함에도 불구하고 테스트가 깨질 수 있지 않을까요?

donggi-lee-bit replied:
다른 테스트로 인해 해당 테스트의 독립성이 깨지는 경우를 방지하기 위해 `@BeforeEach` 절에서 테스트 데이터를 비우는 로직을 추가하겠습니다!

f-lab-maverick commented on src/test/java/com/donggi/sendzy/member/domain/MemberTest.java:
👍

f-lab-maverick commented on src/test/java/com/donggi/sendzy/member/integration/SignupIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+                .then()
+                    .statusCode(HttpStatus.OK.value());
+            }
+        }
```
`RestAssured`를 사용하며 RandomPort를 사용할경우 테스트가끝난 후 데이터가 롤백되지 않으니 주의해주세요~
테스트용 1회성 인메모리 DB를 쓴다면 이슈가 없지만, 개발DB에 연결해서 사용할 경우 테스트데이터가 삭제되지 않습니다.

f-lab-maverick reviewed (changes requested):
꽤나 퀄리티가 좋은 소스코드였습니다 👍 
리뷰 확인하신 후 수정하고, 다시 코드리뷰를 요청해주세요.
고생하셨습니다.

donggi-lee-bit committed:
Lombok을 추가하여 보일러 플레이트 제거

- Lombok 의존성 추가
- @Getter, @RequiredArgsConstructor와 같은 어노테이션을 추가하여 보일러 플레이트 제거

donggi-lee-bit committed:
회원 저장 시 Model 클래스를 도메인 서비스로 넘기도록 변경

donggi-lee-bit committed:
최소, 최대 길이를 관리하는 클래스를 추가하여 Length 변수 선언 방식 개선

donggi-lee-bit committed:
Spring 어노테이션 역할을 주석으로 명시

donggi-lee-bit committed:
MyBatis 쿼리를 xml 파일로 작성

donggi-lee-bit committed:
평문은 plainPassword, 암호화된 필드를 encryptedPassword 로 네이밍 변경

donggi-lee-bit committed:
DataIntegrityViolationException 예외 처리

donggi-lee-bit committed:
MemberRepositoryTest 테스트 메서드가 실행되기 전 BeforeEach를 통해 테스트 데이터 정리 및 TestUtil 에서 rawPassword와 encodedPassword를 구분해서 관리

donggi-lee-bit committed:
RegexPattern enum 클래스에 @Getter 적용

donggi-lee-bit commented:
리뷰해주신 내용 답변 완료하였습니다! 다시 리뷰 요청드립니다. @f-lab-maverick

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/domain/Length.java:
```diff
@@ -0,0 +1,38 @@
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private int min;
+        private int max;
+
+        public Builder min(final int min) {
+            this.min = min;
+            return this;
+        }
+
+        public Builder max(final int max) {
+            this.max = max;
+            return this;
+        }
+
+        public Length build() {
+            return new Length(this);
+        }
```
이 경우 min / max중 하나를 설정하지 않으면 해당 변수에 어떤 값이 들어가게될까요?

donggi-lee-bit replied:
테스트 해 본 결과 min / max 중 하나를 설정하지 않으면 해당 변수는 0으로 초기화됩니다. 이는 java에서 primitive type 변수를 명시적으로 초기화하지 않을 경우 기본 값으로 초기화되기 때문입니다. 만약 Builder에 Reference Type의 변수가 있다면, 해당 변수를 명시적으로 초기화하지 않을 경우 null로 초기화됩니다.

f-lab-maverick commented on src/main/java/com/donggi/sendzy/member/domain/MemberRepository.java:
```diff
@@ -0,0 +1,23 @@
+
+public interface MemberRepository {
+
+    /**
+     * 회원을 저장합니다.
+     * @param member 저장할 회원
+     * @return 저장된 행의 수
+     */
+    int save(Member member);
```
insert의 경우 명확하게 `create` / update의 경우 `update` / `modify`등의 네이밍을 해주시는게 좋겠습니다.
save는 관용적으로 생성 또는 저장이라는 뜻을 겸하여 사용하기 때문에, 다른 개발자들이 보았을 때 헷갈릴 수 있어 주의가 필요합니다.

그래서 JPA에서의 save는 insert와 update를 동시에 하고 있기도 합니다.

donggi-lee-bit replied:
insert는 JPA의 save 메서드가 익숙해서 관용적으로 save를 사용한 것 같습니다. 정확한 의미가 전달되도록 `create` 로 수정하겠습니다!

f-lab-maverick reviewed (changes requested):
코멘트에 답변 및 수정 후 리뷰요청 다시 부탁드립니다~ 고생하셨어요.

donggi-lee-bit committed:
Length 빌더를 제거, 패스워드를 입력된 건 raw, 암호화된 건 encoded 로 구분하여 관리

donggi-lee-bit committed:
Member의 email, encodedPassword 필드의 최소, 최대 길이 변경

- email 은 5자 이상 300자 이하
- encodedPassword 는 60자 이상 300자 이하

donggi-lee-bit committed:
Member insert 메서드 네이밍을 create 로 변경

f-lab-maverick reviewed (approved):
고생하셨습니다. 머지해주셔도 좋습니다~
